### PR TITLE
token lending - bug fixes and flash loan receiver

### PR DESCRIPTION
**Pull request by ilmoi**
Related to [this PR](https://github.com/solana-labs/solana-program-library/pull/2250), but without the CLI and with comments processed.

@jordansexton let me know if you don't want the receiver program included. You didn't comment on it last time so I kept it, but happy to remove.

Once you're happy with this PR I'll issue a second one with the CLI.

### CODE REVIEW

File: token-lending/flash_loan_receiver/src/processor.rs [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        .get(..8)
+        .and_then(|slice| slice.try_into().ok())
+        .map(u64::from_le_bytes)
+        .ok_or(ProgramError::InvalidInstructionData)
+        .map_err(|e| {
+            msg!("Failed to unpack amount.");
+            e
+        })?;
+    Ok(amount)
+}
```
Comment (jordansexton) [inline token-lending/flash_loan_receiver/src/processor.rs:68]: Looks great. I like that you included this as a reference implementation. This should help anyone who wants to use the instruction effectively.

File: token-lending/js/src/instructions/flashLoan.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
+import { TOKEN_PROGRAM_ID } from '@solana/spl-token';
+import { PublicKey, SYSVAR_CLOCK_PUBKEY, TransactionInstruction } from '@solana/web3.js';
```
Comment (jordansexton) [inline token-lending/js/src/instructions/flashLoan.ts:2]: Getting a warning that `SYSVAR_CLOCK_PUBKEY` isn't used.

File: token-lending/js/src/util/layout.ts [hunk 3]
```diff
@@ -79,7 +79,7 @@ export const bigInt =
 
         bigIntLayout.decode = (buffer: Buffer, offset: number) => {
             const src = decode(buffer, offset);
-            return toBigIntLE(src);
+            return toBigIntLE(src as Buffer);
```
Comment (jordansexton) [inline token-lending/js/src/util/layout.ts:82]: Is this still required after the CLI tsconfig change? Mostly wondering if I got something with the types wrong originally.

File: token-lending/program/src/instruction.rs [hunk 4]
```diff
@@ -57,11 +57,11 @@ pub enum LendingInstruction {
     ///   3. `[]` Reserve liquidity SPL Token mint.
     ///   4. `[writable]` Reserve liquidity supply SPL Token account - uninitialized.
     ///   5. `[writable]` Reserve liquidity fee receiver - uninitialized.
-    ///   6. `[]` Pyth product account.
-    ///   7. `[]` Pyth price account.
+    ///   6. `[writable]` Reserve collateral SPL Token mint - uninitialized.
+    ///   7. `[writable]` Reserve collateral token supply - uninitialized.
+    ///   8. `[]` Pyth product account.
+    ///   9. `[]` Pyth price account.
```
Comment (jordansexton) [inline token-lending/program/src/instruction.rs:63]: 👍

**Review by jordansexton — DISMISSED** [review]
Looks good, I only had a few minor questions.

File: token-lending/js/src/instructions/flashLoan.ts (hunk 2 above)
Comment (ilmoi) [reply token-lending/js/src/instructions/flashLoan.ts:2]: My bad, missed it.

File: token-lending/js/src/util/layout.ts (hunk 3 above)
Comment (ilmoi) [reply token-lending/js/src/util/layout.ts:82]: Yep, still getting the error. The full version is:
```
/Users/ilmoi/.nvm/versions/node/v16.1.0/bin/node /Users/ilmoi/Downloads/FORKED_SPL2/token-lending/js/node_modules/ts-node/dist/bin.js /Users/ilmoi/Downloads/FORKED_SPL2/token-lending/js/cli/main.ts
/Users/ilmoi/Downloads/FORKED_SPL2/token-lending/js/node_modules/ts-node/src/index.ts:693
    return new TSError(diagnosticText, diagnosticCodes);
           ^
TSError: ⨯ Unable to compile TypeScript:
../src/util/layout.ts:82:31 - error TS2345: Argument of type 'unknown' is not assignable to parameter of type 'Buffer'.
  Type '{}' is missing the following properties from type 'Buffer': write, toJSON, equals, compare, and 78 more.

82             return toBigIntLE(src);
                                 ~~~

    at createTSError (/Users/ilmoi/Downloads/FORKED_SPL2/token-lending/js/node_modules/ts-node/src/index.ts:693:12)
    at reportTSError (/Users/ilmoi/Downloads/FORKED_SPL2/token-lending/js/node_modules/ts-node/src/index.ts:697:19)
    at getOutput (/Users/ilmoi/Downloads/FORKED_SPL2/token-lending/js/node_modules/ts-node/src/index.ts:884:36)
    at Object.compile (/Users/ilmoi/Downloads/FORKED_SPL2/token-lending/js/node_modules/ts-node/src/index.ts:1186:30)
    at Module.m._compile (/Users/ilmoi/Downloads/FORKED_SPL2/token-lending/js/node_modules/ts-node/src/index.ts:1310:30)
    at Module._extensions..js (node:internal/modules/cjs/loader:1138:10)
    at Object.require.extensions.<computed> [as .ts] (/Users/ilmoi/Downloads/FORKED_SPL2/token-lending/js/node_modules/ts-node/src/index.ts:1314:12)
    at Module.load (node:internal/modules/cjs/loader:989:32)
    at Function.Module._load (node:internal/modules/cjs/loader:829:14)
    at Module.require (node:internal/modules/cjs/loader:1013:19) {
  diagnosticText: "\x1B[96m../src/util/layout.ts\x1B[0m:\x1B[93m82\x1B[0m:\x1B[93m31\x1B[0m - \x1B[91merror\x1B[0m\x1B[90m TS2345: \x1B[0mArgument of type 'unknown' is not assignable to parameter of type 'Buffer'.\n" +
    "  Type '{}' is missing the following properties from type 'Buffer': write, toJSON, equals, compare, and 78 more.\n" +
    '\n' +
    '\x1B[7m82\x1B[0m             return toBigIntLE(src);\n' +
    '\x1B[7m  \x1B[0m \x1B[91m                              ~~~\x1B[0m\n',
  diagnosticCodes: [ 2345 ]
}

Process finished with exit code 1
```

**Comment by jordansexton** [comment]
Merged, thank you!

**Pull request merged**